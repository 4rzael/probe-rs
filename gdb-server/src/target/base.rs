use super::{GdbErrorExt, RuntimeTarget};
use crate::arch::{RuntimeRegId, RuntimeRegisters};

use gdbstub::common::Tid;
use gdbstub::target::ext::base::multithread::MultiThreadBase;
use gdbstub::target::ext::base::multithread::MultiThreadResumeOps;
use gdbstub::target::ext::base::single_register_access::SingleRegisterAccess;
use gdbstub::target::ext::base::single_register_access::SingleRegisterAccessOps;
use gdbstub::target::TargetError;
use probe_rs::{Core, CoreType, InstructionSet, MemoryInterface, RegisterId};

impl MultiThreadBase for RuntimeTarget<'_> {
    fn read_registers(
        &mut self,
        regs: &mut RuntimeRegisters,
        tid: Tid,
    ) -> gdbstub::target::TargetResult<(), Self> {
        let mut session = self.session.borrow_mut();
        let mut core = session.core(tid.get() - 1).into_target_result()?;

        regs.pc = core
            .read_core_reg(core.registers().program_counter())
            .into_target_result()?;

        let mut reg_buffer = Vec::<u8>::new();

        for reg in 0..num_general_registers(&mut core) {
            let (probe_rs_number, bytesize) =
                translate_gdb_register_number(&mut core, reg as u32).unwrap();

            let mut value: u64 = core.read_core_reg(probe_rs_number).unwrap();

            for _ in 0..bytesize {
                let byte = value as u8;
                reg_buffer.push(byte);
                value >>= 8;
            }
        }

        regs.regs = reg_buffer;

        Ok(())
    }

    fn write_registers(
        &mut self,
        regs: &RuntimeRegisters,
        tid: Tid,
    ) -> gdbstub::target::TargetResult<(), Self> {
        let mut session = self.session.borrow_mut();
        let mut core = session.core(tid.get() - 1).into_target_result()?;

        core.write_core_reg(core.registers().program_counter().into(), regs.pc)
            .into_target_result()?;

        let mut current_regval_offset = 0;

        for reg_num in 0..num_general_registers(&mut core) as u32 {
            let (addr, bytesize) = translate_gdb_register_number(&mut core, reg_num).unwrap();

            let current_regval_end = current_regval_offset + bytesize as usize;

            if current_regval_end > regs.regs.len() {
                // Supplied write general registers command argument length not valid, tell GDB
                log::error!(
                    "Unable to write register {}, because supplied register value length was too short",
                    reg_num
                );
                return Err(TargetError::Errno(22));
            }

            let str_value = &regs.regs[current_regval_offset..current_regval_end];

            let mut value = 0;
            for (exp, ch) in str_value.iter().enumerate() {
                value += (*ch as u64) << (8 * exp);
            }

            core.write_core_reg(addr, value).into_target_result()?;

            current_regval_offset = current_regval_end;

            if current_regval_offset == regs.regs.len() {
                break;
            }
        }

        Ok(())
    }

    fn read_addrs(
        &mut self,
        start_addr: u64,
        data: &mut [u8],
        tid: Tid,
    ) -> gdbstub::target::TargetResult<(), Self> {
        let mut session = self.session.borrow_mut();
        let mut core = session.core(tid.get() - 1).into_target_result()?;

        core.read(start_addr, data).into_target_result_non_fatal()
    }

    fn write_addrs(
        &mut self,
        start_addr: u64,
        data: &[u8],
        tid: Tid,
    ) -> gdbstub::target::TargetResult<(), Self> {
        let mut session = self.session.borrow_mut();
        let mut core = session.core(tid.get() - 1).into_target_result()?;

        core.write_8(start_addr, data)
            .into_target_result_non_fatal()
    }

    fn list_active_threads(
        &mut self,
        thread_is_active: &mut dyn FnMut(Tid),
    ) -> Result<(), Self::Error> {
        for i in &self.cores {
            // Unwrap is always safe because we'll never pass 0 to new
            let tid = Tid::new(i + 1).unwrap();
            thread_is_active(tid);
        }

        Ok(())
    }

    fn support_resume(&mut self) -> Option<MultiThreadResumeOps<'_, Self>> {
        Some(self)
    }

    fn support_single_register_access(&mut self) -> Option<SingleRegisterAccessOps<'_, Tid, Self>> {
        Some(self)
    }
}

impl SingleRegisterAccess<Tid> for RuntimeTarget<'_> {
    fn read_register(
        &mut self,
        tid: Tid,
        reg_id: RuntimeRegId,
        buf: &mut [u8],
    ) -> gdbstub::target::TargetResult<usize, Self> {
        let mut session = self.session.borrow_mut();
        let mut core = session.core(tid.get() - 1).into_target_result()?;

        let (probe_rs_number, bytesize) =
            translate_gdb_register_number(&mut core, reg_id.into()).unwrap();

        let mut value: u64 = core.read_core_reg(probe_rs_number).unwrap();

        for i in 0..bytesize {
            let byte = value as u8;
            buf[i as usize] = byte;
            value >>= 8;
        }

        Ok(bytesize as usize)
    }

    fn write_register(
        &mut self,
        tid: Tid,
        reg_id: RuntimeRegId,
        val: &[u8],
    ) -> gdbstub::target::TargetResult<(), Self> {
        let mut session = self.session.borrow_mut();
        let mut core = session.core(tid.get() - 1).into_target_result()?;

        let (probe_rs_number, bytesize) =
            translate_gdb_register_number(&mut core, reg_id.into()).unwrap();

        let mut value = 0;

        for (exp, ch) in val.iter().enumerate().take(bytesize as usize) {
            value += (*ch as u64) << (8 * exp);
        }

        core.write_core_reg(probe_rs_number, value)
            .into_target_result()?;

        Ok(())
    }
}

/// Take a GDB register number and transmate it into a Probe-RS register number
/// for use with [Core::read_core_reg()] and [Core::write_core_reg()]
fn translate_gdb_register_number(
    core: &mut Core,
    gdb_reg_number: u32,
) -> Option<(RegisterId, u32)> {
    let (probe_rs_number, bytesize): (u16, _) = match core.architecture() {
        probe_rs::Architecture::Arm => {
            match core.instruction_set().unwrap_or(InstructionSet::Thumb2) {
                InstructionSet::A64 => match gdb_reg_number {
                    // x0-30, SP, PC
                    x @ 0..=32 => (x as u16, 8),
                    // CPSR
                    x @ 33 => (x as u16, 4),
                    // FPSR
                    x @ 66 => (x as u16, 4),
                    // FPCR
                    x @ 67 => (x as u16, 4),
                    other => {
                        log::warn!("Request for unsupported register with number {}", other);
                        return None;
                    }
                },
                _ => match gdb_reg_number {
                    // Default ARM register (arm-m-profile.xml)
                    // Register 0 to 15
                    x @ 0..=15 => (x as u16, 4),
                    // CPSR register has number 16 in probe-rs
                    // See REGSEL bits, DCRSR register, ARM Reference Manual
                    25 => (16, 4),
                    // Floating Point registers (arm-m-profile-with-fpa.xml)
                    // f0 -f7 start at offset 0x40
                    // See REGSEL bits, DCRSR register, ARM Reference Manual
                    reg @ 16..=23 => ((reg as u16 - 16 + 0x40), 12),
                    // FPSCR has number 0x21 in probe-rs
                    // See REGSEL bits, DCRSR register, ARM Reference Manual
                    24 => (0x21, 4),
                    // Other registers are currently not supported,
                    // they are not listed in the xml files in GDB
                    other => {
                        log::warn!("Request for unsupported register with number {}", other);
                        return None;
                    }
                },
            }
        }
        probe_rs::Architecture::Riscv => match gdb_reg_number {
            // general purpose registers 0 to 31
            x @ 0..=31 => {
                let addr: RegisterId = core
                    .registers()
                    .get_platform_register(x as usize)
                    .expect("riscv register must exist")
                    .into();
                (addr.0, 8)
            }
            // Program counter
            32 => {
                let addr: RegisterId = core.registers().program_counter().into();
                (addr.0, 8)
            }
            other => {
                log::warn!("Request for unsupported register with number {}", other);
                return None;
            }
        },
    };

    Some((RegisterId(probe_rs_number as u16), bytesize))
}

fn num_general_registers(core: &mut Core) -> usize {
    match core.architecture() {
        probe_rs::Architecture::Arm => {
            match core.core_type() {
                // 16 general purpose regs
                CoreType::Armv7a => 16,
                // When in 64 bit mode, 31 GP regs, otherwise 16
                CoreType::Armv8a => {
                    match core.instruction_set().unwrap_or(InstructionSet::Thumb2) {
                        InstructionSet::A64 => 31,
                        _ => 16,
                    }
                }
                // 16 general purpose regs, 8 FP regs
                _ => 24,
            }
        }
        probe_rs::Architecture::Riscv => 33,
    }
}
