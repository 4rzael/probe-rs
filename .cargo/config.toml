[alias]
scripts = "run --package xtask --"
