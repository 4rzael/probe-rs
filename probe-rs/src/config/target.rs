use super::chip::Chip;
use super::flash_algorithm::RawFlashAlgorithm;
use super::memory::MemoryRegion;
use crate::core::{Architecture, CoreType};

/// This describes a complete target with a fixed chip model and variant.
#[derive(Clone)]
pub struct Target {
    /// The name of the target.
    pub name: String,
    /// The name of the flash algorithm.
    pub flash_algorithms: Vec<RawFlashAlgorithm>,
    /// The core type.
    pub core_type: CoreType,
    /// The memory map of the target.
    pub memory_map: Vec<MemoryRegion>,

    /// Source of the target description. Used for diagnostics.
    pub(crate) source: TargetDescriptionSource,
}

impl std::fmt::Debug for Target {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "Target {{
            identifier: {:?},
            flash_algorithms: {:?},
            memory_map: {:?},
        }}",
            self.name, self.flash_algorithms, self.memory_map
        )
    }
}

/// Source of a target description.
///
/// This is used for diagnostics, when
/// an error related to a target description occurs.
#[derive(Clone, Debug, PartialEq)]
pub enum TargetDescriptionSource {
    /// The target description is a generic target description,
    /// which just describes a core type (e.g. M4), without any
    /// flash algorithm or memory description.
    Generic,
    /// The target description is a built-in target description,
    /// which was included into probe-rs at compile time.
    BuiltIn,
    /// The target description was from an external source
    /// during runtime.
    External,
}

/// An error occured while parsing the target description.
pub type TargetParseError = serde_yaml::Error;

impl Target {
    /// Create a new target
    pub fn new(
        chip: &Chip,
        flash_algorithms: Vec<RawFlashAlgorithm>,
        core_type: CoreType,
        source: TargetDescriptionSource,
    ) -> Target {
        Target {
            name: chip.name.clone().into_owned(),
            flash_algorithms,
            core_type,
            memory_map: chip.memory_map.clone().into_owned(),
            source,
        }
    }

    /// Get the architectre of the target
    pub fn architecture(&self) -> Architecture {
        match &self.core_type {
            CoreType::M0 => Architecture::Arm,
            CoreType::M3 => Architecture::Arm,
            CoreType::M33 => Architecture::Arm,
            CoreType::M4 => Architecture::Arm,
            CoreType::M7 => Architecture::Arm,
            CoreType::Riscv => Architecture::Riscv,
        }
    }

    pub fn source(&self) -> &TargetDescriptionSource {
        &self.source
    }
}

/// Selector for the debug target.
#[derive(Debug, Clone)]
pub enum TargetSelector {
    /// Specify the name of a target, which will
    /// be used to search the internal list of
    /// targets.
    Unspecified(String),
    /// Directly specify a target.
    Specified(Target),
    /// Try to automatically identify the target,
    /// by reading identifying information from
    /// the probe and / or target.
    Auto,
}

impl From<&str> for TargetSelector {
    fn from(value: &str) -> Self {
        TargetSelector::Unspecified(value.into())
    }
}

impl From<&String> for TargetSelector {
    fn from(value: &String) -> Self {
        TargetSelector::Unspecified(value.into())
    }
}

impl From<String> for TargetSelector {
    fn from(value: String) -> Self {
        TargetSelector::Unspecified(value)
    }
}

impl From<()> for TargetSelector {
    fn from(_value: ()) -> Self {
        TargetSelector::Auto
    }
}

impl From<Target> for TargetSelector {
    fn from(target: Target) -> Self {
        TargetSelector::Specified(target)
    }
}
