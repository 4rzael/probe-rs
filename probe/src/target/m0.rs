use super::{
    TargetRegister,
    CoreRegisterAddress,
};
use bitfield::bitfield;

bitfield!{
    #[derive(Copy, Clone)]
    pub struct Dhcsr(u32);
    impl Debug;
    pub S_RESET_ST, _: 25;
    pub S_RETIRE_ST, _: 24;
    pub S_LOCKUP, _: 19;
    pub S_SLEEP, _: 18;
    pub S_HALT, _: 17;
    pub S_REGRDY, _: 16;
    pub _, set_C_MASKINTS: 3;
    pub _, set_C_STEP: 2;
    pub _, set_C_HALT: 1;
    pub _, set_C_DEBUGEN: 0;
}

impl From<u32> for Dhcsr {
    fn from(value: u32) -> Self {
        Self(value)
    }
}

impl From<Dhcsr> for u32 {
    fn from(value: Dhcsr) -> Self {
        value.0
    }
}

impl TargetRegister for Dhcsr {
    const ADDRESS: u32 = 0xE000_EDF0;
    const NAME: &'static str = "DHCSR";
}

bitfield!{
    #[derive(Copy, Clone)]
    pub struct Dcrsr(u32);
    impl Debug;
    pub _, set_REGWnR: 16;
    pub _, set_regsel: 4,0;
}

impl From<u32> for Dcrsr {
    fn from(value: u32) -> Self {
        Self(value)
    }
}

impl From<Dcrsr> for u32 {
    fn from(value: Dcrsr) -> Self {
        value.0
    }
}

impl TargetRegister for Dcrsr {
    const ADDRESS: u32 = 0xE000_EDF4;
    const NAME: &'static str = "DCRSR";
}

#[derive(Debug, Copy, Clone)]
pub struct Dcrdr(u32);

impl From<u32> for Dcrdr {
    fn from(value: u32) -> Self {
        Self(value)
    }
}

impl From<Dcrdr> for u32 {
    fn from(value: Dcrdr) -> Self {
        value.0
    }
}

impl TargetRegister for Dcrdr {
    const ADDRESS: u32 = 0xE000_EDF8;
    const NAME: &'static str = "DCRDR";
}

pub const PC: CoreRegisterAddress = CoreRegisterAddress(0b01111);