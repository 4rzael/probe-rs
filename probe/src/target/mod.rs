use crate::flash_writer::FlashAlgorithm;
use crate::debug_probe::{
    MasterProbe,
    DebugProbeError,
    CpuInformation,
};

pub mod m0;

pub trait TargetRegister: Clone + From<u32> + Into<u32> + Sized + std::fmt::Debug {
    const ADDRESS: u32;
    const NAME: &'static str;
}

pub struct CoreRegisterAddress(u8);

impl From<CoreRegisterAddress> for u32 {
    fn from(value: CoreRegisterAddress) -> Self {
        u32::from(value.0)
    }
}

pub struct BasicRegisterAddresses {
    pub R0: CoreRegisterAddress,
    pub R1: CoreRegisterAddress,
    pub R2: CoreRegisterAddress,
    pub R3: CoreRegisterAddress,
    pub R9: CoreRegisterAddress,
    pub PC: CoreRegisterAddress,
    pub LR: CoreRegisterAddress,
    pub SP: CoreRegisterAddress,
}

pub trait Target {
    fn get_flash_algorithm(&self) -> FlashAlgorithm;

    fn get_basic_register_addresses(&self) -> BasicRegisterAddresses;

    fn wait_for_core_halted(&self, mi: &mut MasterProbe) -> Result<(), DebugProbeError>;

    fn halt(&self, mi: &mut MasterProbe) -> Result<CpuInformation, DebugProbeError>;

    fn run(&self, mi: &mut MasterProbe) -> Result<(), DebugProbeError>;

    /// Steps one instruction and then enters halted state again.
    fn step(&self, mi: &mut MasterProbe) -> Result<CpuInformation, DebugProbeError>;

    fn read_core_reg(&self, mi: &mut MasterProbe, addr: CoreRegisterAddress) -> Result<u32, DebugProbeError>;

    fn write_core_reg(&self, mi: &mut MasterProbe, addr: CoreRegisterAddress, value: u32) -> Result<(), DebugProbeError>;

    fn get_available_breakpoint_units(&self, mi: &mut MasterProbe) -> Result<u32, DebugProbeError>;

    fn enable_breakpoints(&self, mi: &mut MasterProbe, state: bool) -> Result<(), DebugProbeError>;

    fn set_breakpoint(&self, mi: &mut MasterProbe, addr: u32) -> Result<(), DebugProbeError>;

    fn enable_breakpoint(&self, mi: &mut MasterProbe, addr: u32) -> Result<(), DebugProbeError>;

    fn disable_breakpoint(&self, mi: &mut MasterProbe, addr: u32) -> Result<(), DebugProbeError>;
}