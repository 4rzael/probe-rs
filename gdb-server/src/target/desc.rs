use super::{GdbErrorExt, RuntimeTarget};

use anyhow::anyhow;

use gdbstub::target::ext::memory_map::MemoryMap;
use gdbstub::target::ext::target_description_xml_override::TargetDescriptionXmlOverride;
use gdbstub::target::TargetError;

use probe_rs::config::MemoryRegion;
use probe_rs::{CoreType, InstructionSet, Session};

fn copy_to_buf(data: &[u8], buf: &mut [u8]) -> usize {
    let len = data.len();
    let buf = &mut buf[..len];
    buf.copy_from_slice(data);
    len
}

fn copy_range_to_buf(data: &[u8], offset: u64, length: usize, buf: &mut [u8]) -> usize {
    let offset = match usize::try_from(offset) {
        Ok(v) => v,
        Err(_) => return 0,
    };
    let len = data.len();
    let data = &data[len.min(offset)..len.min(offset + length)];
    copy_to_buf(data, buf)
}

impl TargetDescriptionXmlOverride for RuntimeTarget<'_> {
    fn target_description_xml(
        &self,
        annex: &[u8],
        offset: u64,
        length: usize,
        buf: &mut [u8],
    ) -> gdbstub::target::TargetResult<usize, Self> {
        let annex = String::from_utf8_lossy(annex);
        if annex != "target.xml" {
            return Err(TargetError::Fatal(
                anyhow!("Unsupported annex: '{}'", annex).into(),
            ));
        }

        let mut session = self.session.borrow_mut();
        let mut core = session.core(self.cores[0]).into_target_result()?;

        let xml = build_target_description(
            core.core_type(),
            core.instruction_set().into_target_result()?,
        );
        let xml_data = xml.as_bytes();

        Ok(copy_range_to_buf(xml_data, offset, length, buf))
    }
}

impl MemoryMap for RuntimeTarget<'_> {
    fn memory_map_xml(
        &self,
        offset: u64,
        length: usize,
        buf: &mut [u8],
    ) -> gdbstub::target::TargetResult<usize, Self> {
        let mut session = self.session.borrow_mut();
        let xml = gdb_memory_map(&mut session, self.cores[0]).into_target_result()?;
        let xml_data = xml.as_bytes();

        Ok(copy_range_to_buf(xml_data, offset, length, buf))
    }
}

/// Compute GDB memory map for a session and primary core
fn gdb_memory_map(
    session: &mut Session,
    primary_core_id: usize,
) -> Result<String, probe_rs::Error> {
    let (virtual_addressing, address_size) = {
        let core = session.core(primary_core_id)?;
        let address_size = core.registers().program_counter().size_in_bits();

        (
            // Cortex-A cores use virtual addressing
            matches!(core.core_type(), CoreType::Armv7a | CoreType::Armv8a),
            address_size,
        )
    };

    let mut xml_map = r#"<?xml version="1.0"?>
<!DOCTYPE memory-map PUBLIC "+//IDN gnu.org//DTD GDB Memory Map V1.0//EN" "http://sourceware.org/gdb/gdb-memory-map.dtd">
<memory-map>
"#.to_owned();

    if virtual_addressing {
        // GDB will not attempt to read / write anything outside the address map.
        // However, with virtual addressing any address could be valid.  As a result
        // we mark the entire address space as RAM since that's the best assumption
        // we can make.
        let region_entry = format!(
            r#"<memory type="ram" start="0x0" length="{:#x}"/>\n"#,
            match address_size {
                32 => 0xFFFF_FFFFu64,
                64 => 0xFFFF_FFFF_FFFF_FFFF,
                _ => 0x0,
            }
        );

        xml_map.push_str(&region_entry);
    } else {
        for region in &session.target().memory_map {
            let region_entry = match region {
                MemoryRegion::Ram(ram) => format!(
                    r#"<memory type="ram" start="{:#x}" length="{:#x}"/>\n"#,
                    ram.range.start,
                    ram.range.end - ram.range.start
                ),
                MemoryRegion::Generic(region) => format!(
                    r#"<memory type="rom" start="{:#x}" length="{:#x}"/>\n"#,
                    region.range.start,
                    region.range.end - region.range.start
                ),
                MemoryRegion::Nvm(region) => {
                    // TODO: Use flash with block size
                    format!(
                        r#"<memory type="rom" start="{:#x}" length="{:#x}"/>\n"#,
                        region.range.start,
                        region.range.end - region.range.start
                    )
                }
            };

            xml_map.push_str(&region_entry);
        }
    }

    xml_map.push_str(r#"</memory-map>"#);

    Ok(xml_map)
}

/// Build the GDB target description XML for a core type and ISA
fn build_target_description(core_type: CoreType, isa: InstructionSet) -> String {
    // GDB-architectures
    //
    // - armv6-m      -> Core-M0
    // - armv7-m      -> Core-M3
    // - armv7e-m      -> Core-M4, Core-M7
    // - armv8-m.base -> Core-M23
    // - armv8-m.main -> Core-M33
    // - riscv:rv32   -> RISCV

    let architecture = match core_type {
        CoreType::Armv6m => "armv6-m",
        CoreType::Armv7a => "armv7",
        CoreType::Armv7m => "armv7",
        CoreType::Armv7em => "armv7e-m",
        CoreType::Armv8a => match isa {
            InstructionSet::A64 => "aarch64",
            _ => "armv8-a",
        },
        CoreType::Armv8m => "armv8-m.main",
        CoreType::Riscv => "riscv:rv32",
    };

    // Only target.xml is supported
    let mut target_description = r#"<?xml version="1.0"?>
        <!DOCTYPE target SYSTEM "gdb-target.dtd">
        <target version="1.0">
        "#
    .to_owned();

    target_description.push_str(&format!("<architecture>{}</architecture>", architecture));

    target_description.push_str("</target>");

    target_description
}

#[cfg(test)]
mod test {
    use super::{build_target_description, CoreType, InstructionSet};

    #[test]
    fn test_target_description_microbit() {
        let description = build_target_description(CoreType::Armv6m, InstructionSet::Thumb2);

        insta::assert_snapshot!(description);
    }
}
