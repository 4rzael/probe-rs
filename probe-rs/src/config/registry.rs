//! Internal target registry

use super::target::{Target, TargetDescriptionSource};
use crate::config::{Chip, ChipFamily, ChipInfo};
use crate::core::CoreType;
use lazy_static::lazy_static;
use std::fs::File;
use std::path::Path;
use std::{
    borrow::Cow,
    sync::{Arc, Mutex, TryLockError},
};
use thiserror::Error;

lazy_static! {
    static ref REGISTRY: Arc<Mutex<Registry>> =
        Arc::new(Mutex::new(Registry::from_builtin_families()));
}

/// Error type for all errors which occur when working
/// with the internal registry of targets.
#[derive(Debug, Error)]
pub enum RegistryError {
    /// The requested chip was not found in the registry.
    #[error("The requested chip '{0}' was not found in the list of known targets.")]
    ChipNotFound(String),
    /// When searching for a chip based on information read from the target,
    /// no matching chip was found in the registry.
    #[error("The connected chip could not automatically be determined.")]
    ChipAutodetectFailed,
    /// A core type contained in a target description is not supported
    /// in probe-rs.
    #[error("The core type '{0}' is not supported in probe-rs.")]
    UnknownCoreType(String),
    /// An IO error which occured when trying to read a target description file.
    #[error("An IO error was encountered")]
    Io(#[from] std::io::Error),
    /// An error occured while deserializing a YAML target description file.
    #[error("Deserializing the yaml encountered an error")]
    Yaml(#[from] serde_yaml::Error),
    /// Unable to lock the registry.
    #[error("Unable to lock registry")]
    LockUnavailable,
}

impl<R> From<TryLockError<R>> for RegistryError {
    fn from(_: TryLockError<R>) -> Self {
        RegistryError::LockUnavailable
    }
}

const GENERIC_TARGETS: [ChipFamily; 6] = [
    ChipFamily {
        name: Cow::Borrowed("Generic Cortex-M0"),
        manufacturer: None,
        variants: Cow::Borrowed(&[Chip {
            name: Cow::Borrowed("cortex-m0"),
            part: None,
            memory_map: Cow::Borrowed(&[]),
            flash_algorithms: Cow::Borrowed(&[]),
        }]),
        flash_algorithms: Cow::Borrowed(&[]),
        core: Cow::Borrowed("M0"),
        source: TargetDescriptionSource::Generic,
    },
    ChipFamily {
        name: Cow::Borrowed("Generic Cortex-M4"),
        manufacturer: None,
        variants: Cow::Borrowed(&[Chip {
            name: Cow::Borrowed("cortex-m4"),
            part: None,
            memory_map: Cow::Borrowed(&[]),
            flash_algorithms: Cow::Borrowed(&[]),
        }]),
        flash_algorithms: Cow::Borrowed(&[]),
        core: Cow::Borrowed("M4"),
        source: TargetDescriptionSource::Generic,
    },
    ChipFamily {
        name: Cow::Borrowed("Generic Cortex-M3"),
        manufacturer: None,
        variants: Cow::Borrowed(&[Chip {
            name: Cow::Borrowed("cortex-m3"),
            part: None,
            memory_map: Cow::Borrowed(&[]),
            flash_algorithms: Cow::Borrowed(&[]),
        }]),
        flash_algorithms: Cow::Borrowed(&[]),
        core: Cow::Borrowed("M3"),
        source: TargetDescriptionSource::Generic,
    },
    ChipFamily {
        name: Cow::Borrowed("Generic Cortex-M33"),
        manufacturer: None,
        variants: Cow::Borrowed(&[Chip {
            name: Cow::Borrowed("cortex-m33"),
            part: None,
            memory_map: Cow::Borrowed(&[]),
            flash_algorithms: Cow::Borrowed(&[]),
        }]),
        flash_algorithms: Cow::Borrowed(&[]),
        core: Cow::Borrowed("M33"),
        source: TargetDescriptionSource::Generic,
    },
    ChipFamily {
        name: Cow::Borrowed("Generic Cortex-M7"),
        manufacturer: None,
        variants: Cow::Borrowed(&[Chip {
            name: Cow::Borrowed("cortex-m7"),
            part: None,
            memory_map: Cow::Borrowed(&[]),
            flash_algorithms: Cow::Borrowed(&[]),
        }]),
        flash_algorithms: Cow::Borrowed(&[]),
        core: Cow::Borrowed("M7"),
        source: TargetDescriptionSource::Generic,
    },
    ChipFamily {
        name: Cow::Borrowed("Generic Riscv"),
        manufacturer: None,
        variants: Cow::Borrowed(&[Chip {
            name: Cow::Borrowed("riscv"),
            part: None,
            memory_map: Cow::Borrowed(&[]),
            flash_algorithms: Cow::Borrowed(&[]),
        }]),
        flash_algorithms: Cow::Borrowed(&[]),
        core: Cow::Borrowed("riscv"),
        source: TargetDescriptionSource::Generic,
    },
];

/// Registry of all available targets.
struct Registry {
    /// All the available chips.
    families: Vec<ChipFamily>,
}

#[cfg(feature = "builtin-targets")]
mod builtin {
    include!(concat!(env!("OUT_DIR"), "/targets.rs"));
}

impl Registry {
    #[cfg(feature = "builtin-targets")]
    fn from_builtin_families() -> Self {
        let mut families = Vec::from(builtin::get_targets());

        families.extend(GENERIC_TARGETS.iter().cloned());

        Self { families }
    }

    #[cfg(not(feature = "builtin-targets"))]
    fn from_builtin_families() -> Self {
        Self {
            families: GENERIC_TARGETS.iter().cloned().collect(),
        }
    }

    fn families(&self) -> &Vec<ChipFamily> {
        &self.families
    }

    fn get_target_by_name(&self, name: impl AsRef<str>) -> Result<Target, RegistryError> {
        let name = name.as_ref();

        log::debug!("Searching registry for chip with name {}", name);

        let (family, chip) = {
            // Try get the corresponding chip.
            let mut selected_family_and_chip = None;
            for family in &self.families {
                for variant in family.variants.iter() {
                    if variant
                        .name
                        .to_ascii_lowercase()
                        .starts_with(&name.to_ascii_lowercase())
                    {
                        if variant.name.to_ascii_lowercase() != name.to_ascii_lowercase() {
                            log::warn!(
                                "Found chip {} which matches given partial name {}. Consider specifying its full name.",
                                variant.name,
                                name,
                            )
                        }
                        selected_family_and_chip = Some((family, variant));
                    }
                }
            }
            let (family, chip) = selected_family_and_chip
                .ok_or_else(|| RegistryError::ChipNotFound(name.to_owned()))?;

            // Try get the correspnding flash algorithm.
            (family, chip)
        };
        self.get_target(family, chip)
    }

    fn get_target_by_chip_info(&self, chip_info: ChipInfo) -> Result<Target, RegistryError> {
        let (family, chip) = {
            match chip_info {
                ChipInfo::Arm(chip_info) => {
                    // Try get the corresponding chip.

                    let families = self.families.iter().filter(|f| {
                        f.manufacturer
                            .map(|m| m == chip_info.manufacturer)
                            .unwrap_or(false)
                    });

                    let mut identified_chips = Vec::new();

                    for family in families {
                        log::debug!("Checking family {}", family.name);

                        let chips = family
                            .variants()
                            .iter()
                            .filter(|v| v.part.map(|p| p == chip_info.part).unwrap_or(false))
                            .map(|c| (family, c));

                        identified_chips.extend(chips)
                    }

                    if identified_chips.len() == 1 {
                        identified_chips.pop().unwrap()
                    } else {
                        log::debug!(
                        "Found {} matching chips for information {:?}, unable to determine chip",
                        identified_chips.len(),
                        chip_info
                    );
                        return Err(RegistryError::ChipAutodetectFailed);
                    }
                }
            }
        };
        self.get_target(family, chip)
    }

    fn get_target(&self, family: &ChipFamily, chip: &Chip) -> Result<Target, RegistryError> {
        // Try get the corresponding chip.
        let core = if let Some(core) = CoreType::from_string(&family.core) {
            core
        } else {
            return Err(RegistryError::UnknownCoreType(
                family.core.clone().into_owned(),
            ));
        };

        // find relevant algorithms
        let chip_algorithms = chip
            .flash_algorithms
            .iter()
            .filter_map(|fa| family.get_algorithm(fa))
            .cloned()
            .collect();

        Ok(Target::new(
            chip,
            chip_algorithms,
            core,
            family.source.clone(),
        ))
    }

    fn add_target_from_yaml(&mut self, path_to_yaml: &Path) -> Result<(), RegistryError> {
        let file = File::open(path_to_yaml)?;
        let chip = ChipFamily::from_yaml_reader(file)?;

        let index = self
            .families
            .iter()
            .position(|old_chip| old_chip.name == chip.name);
        if let Some(index) = index {
            self.families.remove(index);
        }
        self.families.push(chip);

        Ok(())
    }
}

/// Get a target from the internal registry based on its name.
pub fn get_target_by_name(name: impl AsRef<str>) -> Result<Target, RegistryError> {
    REGISTRY.try_lock()?.get_target_by_name(name)
}

/// Try to retrieve a target based on [ChipInfo] read from a target.
pub(crate) fn get_target_by_chip_info(chip_info: ChipInfo) -> Result<Target, RegistryError> {
    REGISTRY.try_lock()?.get_target_by_chip_info(chip_info)
}

/// Parse a target description file and add the contained targets
/// to the internal target registry.
pub fn add_target_from_yaml(path_to_yaml: &Path) -> Result<(), RegistryError> {
    REGISTRY.try_lock()?.add_target_from_yaml(path_to_yaml)
}

/// Get a list of all families which are contained in the internal
/// registry.
pub fn families() -> Result<Vec<ChipFamily>, RegistryError> {
    Ok(REGISTRY.try_lock()?.families().clone())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn try_fetch1() {
        let registry = Registry::from_builtin_families();
        assert!(registry.get_target_by_name("nrf51").is_ok());
    }

    #[test]
    fn try_fetch2() {
        let registry = Registry::from_builtin_families();
        assert!(registry.get_target_by_name("nrf5182").is_ok());
    }

    #[test]
    fn try_fetch3() {
        let registry = Registry::from_builtin_families();
        assert!(registry.get_target_by_name("nrF51822_x").is_ok());
    }

    #[test]
    fn try_fetch4() {
        let registry = Registry::from_builtin_families();
        assert!(registry.get_target_by_name("nrf51822_Xxaa").is_ok());
    }
}
