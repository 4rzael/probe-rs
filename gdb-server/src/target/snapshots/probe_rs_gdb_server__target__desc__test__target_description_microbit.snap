---
source: gdb-server/src/target/desc.rs
assertion_line: 186
expression: description
---
<?xml version="1.0"?>
        <!DOCTYPE target SYSTEM "gdb-target.dtd">
        <target version="1.0">
        <architecture>armv6-m</architecture></target>
